docs-internal/message-body-api.md

```diff
@@ -0,0 +1,918 @@
+
+This event mechanism is part of the public API, however it is more of an internal API ; the end-user exposure is very minimal.
+
+```java
+public interface EventListener {
+    void onEvent(Event event);
+}
+
+public enum EventType {
+    BEFORE_ONSUBSCRIBE,
+    AFTER_ONSUBSCRIBE,
+    BEFORE_ONNEXT,
+    AFTER_ONNEXT,
+    BEFORE_ONERROR,
+    AFTER_ONERROR,
+    BEFORE_ONCOMPLETE,
+    AFTER_ONCOMPLETE
+}
+
+public interface Event {
+
+    EventType eventType();
+    Optional<GenericType<?>> entityType();
+}
+
+public interface ErrorEvent extends Event {
+    Throwable error();
+}
+```
+
+### MessageBodyContent
+
+```java
+public interface MessageBodyContent extends Publisher<DataChunk> {
+}
+```
+
+This interface extends `Publisher<DataChunk>` and does not add any other method. It is used to model the "raw" message body, it is a base interface for readable message body and writeable message body. The use of this interface is not forced onto the end users, instead it is only used as part of classes and interfaces definition ; methods consuming or returning "raw" payload use `Publisher<DataChunk>`.
+
+This interface is also used as a place holder to clarify some behavior aspect of the provided publisher implements with Javadocs comment. Such behaviors are: single subscriber and non resumable (i.e cancel the subscription and request more items later on).
+
+### MessageBodyFilter
+
+The new style of filter. It is now a reactive processor instead of a `Function<Publisher<DataChunk>, Publisher<DataChunk>`.
+
+```java
+public interface MessageBodyFilter extends Processor<DataChunk, DataChunk> {
+}
+```
+
+Because of the single subscriber limitation of the Helidon provided publishers, the filters must be registered using a special mechanism. They can now be implemented as reactive processor, but cannot be registered as reactive processor.
+
+### MessageBodyReaderContext
+
+This is a new class that exposes a re-usable configured mechanism to unmarshall the message body "raw" payload into objects with reactive constructs.
+It uses 2 `MessageBodyOperators` registry, one for the regular readers (i.e the registered `MessageBodyReader` instances) and one for the stream readers (i.e the registered `MessageBodyStreamReader` instances).
+
+It provides the following methods:
+
+```java
+@Override
+public MessageBodyReaderContext registerReader(MessageBodyReader<?> reader) {
+  // ...
+}
+
+@Override
+public MessageBodyReaderContext registerReader(MessageBodyStreamReader<?> reader) {
+  // ...
+}
+
+@Deprecated
+public <T> void registerReader(Class<T> type,
+  io.helidon.common.http.Reader<T> reader) {
+  // ...
+}
+
+@Deprecated
+public <T> void registerReader(Predicate<Class<?>> predicate,
+  io.helidon.common.http.Reader<T> reader) {
```

Comment:
Add a note to state that the FQDN notation for `Reader` is to avoid the compiler deprecation warning.